### PR TITLE
Implement user defined planner for position

### LINKED ISSUE #11242: Implement user defined planner for sql_position_to_expr
### Is your feature request related to a problem or challenge?

Part of #11207

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xinlifoobar**
## Which issue does this PR close?

Closes #11242

### CODE REVIEW

**Comment by alamb** [comment]
🤔 We seem to be hitting the dreaded sql planner stack overflow in debug builds again 

https://github.com/apache/datafusion/actions/runs/9778881627/job/26996930268?pr=11243


```
thread 'tokio-runtime-worker' has overflowed its stack
fatal runtime error: stack overflow
error: test failed, to rerun pass `-p datafusion-sqllogictest --test sqllogictests`

Caused by:
  process didn't exit successfully: `/__w/datafusion/datafusion/target/debug/deps/sqllogictests-dd5da520160d5a7d` (signal: 6, SIGABRT: process abort signal)
Error: Process completed with exit code 101.
```

I think best way to fix this is to make a function (rather than inlining the code into the same big match expression)

File: datafusion/core/src/execution/session_state.rs [hunk 1]
```diff
@@ -967,14 +967,19 @@ impl SessionState {
...
                 .with_user_defined_planner(field_access_planner)
         }
-        #[cfg(not(feature = "array_expressions"))]
+        #[cfg(feature = "unicode_expressions")]
         {
-            query
+            let position_planner =
+                Arc::new(functions::unicode::planner::PositionPlanner::default()) as _;
+
+            query = query.with_user_defined_planner(position_planner);
```
Comment (dharanad) [inline datafusion/core/src/execution/session_state.rs:979]: As part of this issues #https://github.com/apache/datafusion/pull/11253 we have deided to move registration of SQL planner to constructor `pub fn new_with_config_rt_and_catalog_list`

**Review by alamb — APPROVED** [review]
Thank you for the contribution @xinlifoobar  -- this makes sense to me

**Review by comphead — APPROVED** [review]
lgtm thanks @xinlifoobar but I have same concern as @alamb on having a separate planner

File: datafusion/functions/src/unicode/planner.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+use datafusion_expr::{
+    expr::ScalarFunction,
+    planner::{PlannerResult, UserDefinedSQLPlanner},
+    Expr,
+};
+
+#[derive(Default)]
+pub struct PositionPlanner;
+
+impl UserDefinedSQLPlanner for PositionPlanner {
```
[the remark this answers is not in the document]
Comment (alamb) [reply datafusion/functions/src/unicode/planner.rs:30]: Filed https://github.com/apache/datafusion/issues/11305 to track. Thank you

**Comment by alamb** [comment]
🚀  let's consolidate the planners as a follow on PR

**Pull request merged**